### PR TITLE
Implement Serde for fields and curves

### PR DESCRIPTION
**Pull request by vmx**
Currently Serde is only implemented for `Fp`, `Fq`, `EpAffine` and `EqAffine`.
Support can be enabled with the `serde` feature.

It's based on the implementation in `blstrs`: https://github.com/filecoin-project/blstrs/blob/master/src/serde_impl.rs

Please let me know if that's the correct approach, or e.g. you'd like to serialize them to four `u64`s.

I've only implemented it for the things I currently need, but I'm happy to provide an implementation for all curves.

### CODE REVIEW

File: src/serde_impl.rs
Comment (daira) [inline src/serde_impl.rs:188]: I have not checked this value (or the one for `EqAffine::generator()` below).

**Review by daira — APPROVED** [review]
utACK. Serialising the byte representations looks good to me.

File: src/serde_impl.rs
Comment (daira) [inline src/serde_impl.rs:53]:
```suggestion
        match EpAffine::from_bytes(&bytes).into() {
```
If you look at the implementation in https://github.com/zcash/pasta_curves/blob/main/src/curves.rs , `from_bytes_unchecked` and `from_bytes` have exactly the same behaviour, but using `from_bytes` is clearer because in fact the representation is checked.

File: src/serde_impl.rs
Comment (daira) [inline src/serde_impl.rs:69]:
```suggestion
        match EqAffine::from_bytes(&bytes).into() {
```

**Comment by parazyd** [comment]
@vmx Would you also consider adding borsh support or is that out of scope?

**Comment by vmx** [comment]
@parazyd I won't have time working on borsh support.

**Comment by parazyd** [comment]
@vmx Alright, I will then add it after this PR is merged. Thank you.

File: Cargo.toml
Comment (arthurgreef) [inline Cargo.toml:71]: Spelling. Dependency.

**Comment by FrankC01** [comment]
Is this going to be merged in?

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:112]: Serializing a byte encoding as a JSON array of integers is not ideal (both because it is unsuitable for a human-readable format, and significantly more verbose). What we should instead do is [detect whether the encoding format is human readable](https://docs.rs/serde/latest/serde/trait.Serializer.html#method.is_human_readable), and if it is, encode the bytes as a hex string. Then on deserialization we do the [equivalent human-readable check](https://docs.rs/serde/latest/serde/trait.Deserializer.html#method.is_human_readable), and expect to decode either a 64-byte hex string, or a 32-byte array.

File: Cargo.toml
Comment (str4d) [inline Cargo.toml:62]: Does this need to pin `serde 1.0.144`? I would prefer that we only specify the minimum required `serde` version, for maximal compatibility with downstream users. For my prior comment, this would mean a version of `serde` that supports `{Serializer, Deserializer}::is_human_readable`.

**Review by str4d — CHANGES_REQUESTED** [review]

**Comment by str4d** [comment]
This PR needs to be rebased (GitHub's UI indicates there is a merge conflict).

**Comment by vmx** [comment]
As I needed to rebase, I did merged all commits into a single one. It should now include all the things mentioned in the review.

Notable things are:
 - I've also added `bincode` to test the non human readable version.
 - I've kept `serde` as feature name. The "hack" renaming the crate was inspired https://github.com/RustCrypto/RSA/pull/41/files. Once the minimum required Rust version is 1.60, the `dep:serde` can be used in the feature flag definitions

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:79]: The changes addressing @daira's earlier comments were lost during the rebase:
```suggestion
        match EpAffine::from_bytes(&bytes).into() {
```

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:95]:
```suggestion
        match EqAffine::from_bytes(&bytes).into() {
```

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:81]: `ERR_CODE` here is wrong, as this is a Pallas curve point, not a field element.

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:97]: `ERR_CODE` here is wrong, as this is a Vesta curve point, not a field element.

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:49]: It would be preferable for the error to say "a Pallas field element" or "an Fp element".

File: src/serde_impl.rs
Comment (str4d) [inline src/serde_impl.rs:65]: It would be preferable for the error to say "a Vesta field element" or "an Fq element".

**Review by str4d — CHANGES_REQUESTED** [review]
Human-readable serialization change looks good.

**Comment by samuelburnham** [comment]
Following this model, I have also implemented Serde for the `Ep` and `Eq` types. Should I open a separate PR off this one or wait till it's merged?

**Review by str4d — APPROVED** [review]
utACK 025362ad5af96e9ca54851d3554af6ca1d0c2cf6

**Pull request merged**
